File: typecast-c/examples/quick_cloning.c
/**
 * Typecast C SDK - Quick Voice Cloning Example
 *
 * Demonstrates how to clone a custom voice from an audio file and then
 * use it for text-to-speech synthesis.
 *
 * Build: compile with CMake (see CMakeLists.txt)
 *
 * Usage:
 *   ./example_quick_cloning <audio_file> [voice_name]
 *   or set TYPECAST_API_KEY environment variable
 *
 * Example:
 *   TYPECAST_API_KEY=your_key ./example_quick_cloning sample.wav "My Voice"
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "typecast.h"

static const char* get_api_key(int argc, char* argv[]) {
    (void)argc; (void)argv;
    const char* key = getenv("TYPECAST_API_KEY");
    if (key && strlen(key) > 0) return key;
    return NULL;
}

/* Read an entire file into a heap buffer.  Returns NULL on error. */
static unsigned char* read_file(const char* path, size_t* out_len) {
    FILE* f = fopen(path, "rb");
    if (!f) { fprintf(stderr, "Cannot open %s\n", path); return NULL; }

    fseek(f, 0, SEEK_END);
    long size = ftell(f);
    fseek(f, 0, SEEK_SET);

    if (size <= 0 || (size_t)size > (size_t)TYPECAST_CLONING_MAX_FILE_SIZE) {
        fprintf(stderr, "File size out of range (0 < size <= 25 MB)\n");
        fclose(f);
        return NULL;
    }

    unsigned char* buf = (unsigned char*)malloc((size_t)size);
    if (!buf) { fclose(f); return NULL; }

    if (fread(buf, 1, (size_t)size, f) != (size_t)size) {
        free(buf); fclose(f); return NULL;
    }
    fclose(f);
    *out_len = (size_t)size;
    return buf;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <audio_file> [voice_name]\n", argv[0]);
        return 1;
    }

    const char* audio_path = argv[1];
    const char* voice_name = (argc >= 3) ? argv[2] : "My Cloned Voice";

    const char* api_key = get_api_key(argc, argv);
    if (!api_key) {
        fprintf(stderr, "Set TYPECAST_API_KEY environment variable.\n");
        return 1;
    }

    /* --- Create client --- */
    TypecastClient* client = typecast_client_create(api_key);
    if (!client) {
        fprintf(stderr, "Failed to create client.\n");
        return 1;
    }

    /* --- Read audio file --- */
    size_t audio_len = 0;
    unsigned char* audio = read_file(audio_path, &audio_len);
    if (!audio) {
        typecast_client_destroy(client);
        return 1;
    }
    printf("Loaded audio: %s (%zu bytes)\n", audio_path, audio_len);

    /* --- Clone voice --- */
    printf("Cloning voice \"%s\" ...\n", voice_name);
    TypecastCustomVoice custom_voice;
    TypecastErrorCode rc = typecast_clone_voice(
        client,
        audio, audio_len,
        audio_path,       /* filename hint for MIME type detection */
        voice_name,
        "ssfm-v30",
        &custom_voice
    );
    free(audio);

    if (rc != TYPECAST_OK) {
        const TypecastError* err = typecast_client_get_error(client);
        fprintf(stderr, "Clone failed: %s (code %d)\n",
                err && err->message ? err->message : "unknown", rc);
        typecast_client_destroy(client);
        return 1;
    }

    printf("Clone successful!\n");
    printf("  voice_id : %s\n", custom_voice.voice_id);
    printf("  name     : %s\n", custom_voice.name);
    printf("  model    : %s\n", custom_voice.model);

    /* --- Use the cloned voice for TTS --- */
    printf("\nGenerating speech with the cloned voice...\n");

    TypecastTTSRequest tts_req;
    memset(&tts_req, 0, sizeof(tts_req));
    tts_req.text     = "Hello, this is my custom cloned voice speaking!";
    tts_req.voice_id = custom_voice.voice_id;
    tts_req.model    = TYPECAST_MODEL_SSFM_V30;

    TypecastTTSResponse* tts_resp = typecast_text_to_speech(client, &tts_req);
    if (tts_resp) {
        const char* out_path = "cloned_voice_output.wav";
        FILE* f = fopen(out_path, "wb");
        if (f) {
            fwrite(tts_resp->audio_data, 1, tts_resp->audio_size, f);
            fclose(f);
            printf("Audio saved to %s (%.2f seconds)\n",
                   out_path, tts_resp->duration);
        }
        typecast_tts_response_free(tts_resp);
    } else {
        const TypecastError* err = typecast_client_get_error(client);
        fprintf(stderr, "TTS failed: %s\n",
                err && err->message ? err->message : "unknown");
    }

    /* --- Delete the cloned voice --- */
    printf("\nDeleting cloned voice %s ...\n", custom_voice.voice_id);
    rc = typecast_delete_voice(client, custom_voice.voice_id);
    if (rc == TYPECAST_OK) {
        printf("Voice deleted.\n");
    } else {
        const TypecastError* err = typecast_client_get_error(client);
        fprintf(stderr, "Delete failed: %s (code %d)\n",
                err && err->message ? err->message : "unknown", rc);
    }

    typecast_client_destroy(client);
    return 0;
}

File: typecast-rust/examples/quick_cloning.rs
//! Quick Voice Cloning example
//!
//! Clones a voice from a local audio file and optionally synthesizes speech
//! with the new voice.
//!
//! # Usage
//!
//! ```bash
//! # Clone from a WAV file with default model (ssfm-v30)
//! cargo run --example quick_cloning -- sample.wav "My Voice"
//!
//! # Specify model explicitly
//! cargo run --example quick_cloning -- sample.wav "My Voice" ssfm-v21
//! ```
//!
//! Set `TYPECAST_API_KEY` in your environment before running.

use std::env;
use typecast_rust::{TypecastClient, TypecastError};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 3 {
        eprintln!("Usage: {} <audio_file> <voice_name> [model]", args[0]);
        eprintln!("       model defaults to ssfm-v30");
        std::process::exit(1);
    }

    let audio_path = &args[1];
    let voice_name = &args[2];
    let model = args.get(3).map(String::as_str).unwrap_or("ssfm-v30");

    let client = TypecastClient::from_env().map_err(|e| {
        eprintln!("Failed to create client: {e}");
        eprintln!("Make sure TYPECAST_API_KEY is set.");
        e
    })?;

    // Read the audio file.
    let audio = std::fs::read(audio_path).map_err(|e| {
        eprintln!("Could not read audio file '{audio_path}': {e}");
        e
    })?;

    let filename = std::path::Path::new(audio_path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or(audio_path.as_str());

    println!("Cloning voice from '{audio_path}' ({} bytes) ...", audio.len());

    let voice = match client
        .clone_voice(audio, filename, voice_name, model)
        .await
    {
        Ok(v) => v,
        Err(TypecastError::ValidationError { detail }) => {
            eprintln!("Validation error: {detail}");
            std::process::exit(1);
        }
        Err(e) => return Err(e.into()),
    };

    println!("Voice cloned successfully!");
    println!("  voice_id : {}", voice.voice_id);
    println!("  name     : {}", voice.name);
    println!("  model    : {}", voice.model);
    println!();
    println!("To delete this voice later, run:");
    println!(
        "  TYPECAST_API_KEY=... cargo run --example quick_cloning -- delete {}",
        voice.voice_id
    );

    Ok(())
}
